### PR TITLE
REFACTORING elements

### PR DESCRIPTION
**Pull request by petrasvestartas**
Refactoring of elements, 5 classes:

- elements.py https://github.com/BRG-research/compas_model/blob/b7e2aea923804ba6c913b1233f3d575543305ddc/src/compas_model/elements/element.py
- block.py https://github.com/BRG-research/compas_model/blob/b7e2aea923804ba6c913b1233f3d575543305ddc/src/compas_model/elements/zero_dimensional/block.py
- beam.py https://github.com/BRG-research/compas_model/blob/b7e2aea923804ba6c913b1233f3d575543305ddc/src/compas_model/elements/one_dimensional/beam.py
- plate.py https://github.com/BRG-research/compas_model/blob/b7e2aea923804ba6c913b1233f3d575543305ddc/src/compas_model/elements/two_dimensional/plate.py
- interface.py https://github.com/BRG-research/compas_model/blob/b7e2aea923804ba6c913b1233f3d575543305ddc/src/compas_model/elements/two_dimensional/interface.py

Elements

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [x] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas_model/elements/element.py [hunk 1]
```diff
@@ -2,87 +2,320 @@
...
-)
+from compas.geometry import Frame
+from compas.geometry import Vector
+from compas.geometry import Box
+from abc import abstractmethod
 
 
 class Element(Data):
-    """Class representing a structural object.
+    """Template class of an element that has a link with a model class.
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: Base class for model elements.

File: src/compas_model/elements/element.py [hunk 2]
```diff
@@ -2,87 +2,320 @@
...
+    Do not use this class directly; instead, use a subclass such as Beam, Block, Plate, Interface, etc.
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: leave a blank line and add this as a warning

File: src/compas_model/elements/element.py [hunk 3]
```diff
@@ -2,87 +2,320 @@
...
-    >>> element = Element("BLOCK", Frame.worldXY(), Point(0,0,0), box)
-    >>> element.insertion = Vector(0, 0, 1)
+    Attributes
+    ----------
+    dtype : str, read-only
+        The type of the object in the form of a fully qualified module name and a class name, separated by a forward slash ("/").
+        For example: ``"compas.datastructures/Mesh"``.
+    data : dict
+        The representation of the object as a dictionary containing only built-in Python data types.
+        The structure of the dict is described by the data schema.
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: remove

File: src/compas_model/elements/element.py [hunk 4]
```diff
@@ -2,87 +2,320 @@
...
+    name : str
+        The name of the object.
+        This name is not necessarily unique and can be set by the user.
+        The default value is the object's class name: ``self.__class__.__name__``.
+    frame : :class:`compas.geometry.Frame`
+        Local coordinate of the object, default is :class:`compas.geometry.Frame.WorldXY()`.
+    geometry_simplified : Any
+        Minimal geometrical represetation of an object. For example a :class:`compas.geometry.Polyline` that can represent: a point, a line or a polyline.
+    geometry : Any
+        A list of closed shapes. For example a box of a beam, a mesh of a block and etc.
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: is this always a list?

File: src/compas_model/elements/element.py [hunk 5]
```diff
@@ -2,87 +2,320 @@
...
+        The globally unique identifier of the object.
+        The guid is generated with ``uuid.uuid4()``.
+    name : str
+        The name of the object.
+        This name is not necessarily unique and can be set by the user.
+        The default value is the object's class name: ``self.__class__.__name__``.
+    frame : :class:`compas.geometry.Frame`
+        Local coordinate of the object, default is :class:`compas.geometry.Frame.WorldXY()`.
+    geometry_simplified : Any
+        Minimal geometrical represetation of an object. For example a :class:`compas.geometry.Polyline` that can represent: a point, a line or a polyline.
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: i don't understand this description. a polyline that can represent a point, line or polyline?

File: src/compas_model/elements/element.py (hunk 3 above)
Comment (tomvanmele) [reply src/compas_model/elements/element.py]: the data API is no longer public in the latest version of compas

File: src/compas_model/elements/element.py [hunk 6]
```diff
@@ -2,87 +2,320 @@
...
+    aabb : :class:`compas.geometry.Box`
+        The Axis Aligned Bounding Box (AABB) of the element.
+    obb : :class:`compas.geometry.Box`
+        The Oriented Bounding Box (OBB) of the element.
+    collision_mesh : :class:`compas.datastructures.Mesh`
+        The collision geometry of the element.
+    dimensions : list
+        The dimensions of the element.
+    features : dict
+        The features of the element, joinery, openings, etc.
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: i would explain a bit better what the point is of these features. e.g. they are added to the geometry through operations specified by the user?

File: src/compas_model/elements/element.py [hunk 7]
```diff
@@ -2,87 +2,320 @@
...
+    insertion : :class:`compas.geometry.Vector`
+        The insertion vector of the element. Default is (0, 0, -1).
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: again, explain what this is...

File: src/compas_model/elements/element.py [hunk 8]
```diff
@@ -2,87 +2,320 @@
...
+    node : :class:`compas_model.model.ElementNode`
+        The node of the element.
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: the node in the model tree containing the element

File: src/compas_model/elements/element.py [hunk 9]
```diff
@@ -2,87 +2,320 @@
...
 
     def __init__(
-        self,
-        name=None,
-        frame=None,
-        geometry_simplified=None,
-        geometry=None,
-        copy_geometry=True,
-        **kwargs
+        self, name=None, frame=None, geometry_simplified=None, geometry=None, **kwargs
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: i would switch the order of `geometry_simplified` and `geometry`. also in the attributes list. and give an example of what `geometry_simplified` could be in relation to what `geometry` is...

File: src/compas_model/elements/element.py [hunk 10]
```diff
@@ -2,87 +2,320 @@
...
-        # --------------------------------------------------------------------------
-        # Define this property dynamically in the class.
-        # --------------------------------------------------------------------------
+
+        name = name.lower() if name else str.lower(self.__class__.__name__)
+        super(Element, self).__init__(name=name, **kwargs)
+
+        self._frame = frame if frame.copy() else Frame.worldXY()
+        self._geometry_simplified = (
+            self._copy_geometries(geometry_simplified) if geometry else []
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: `if geometry_simplified`?

File: src/compas_model/elements/element.py [hunk 11]
```diff
@@ -2,87 +2,320 @@
...
         self._aabb = None
-        self._oobb = None
-        self._inflate = 0.0
-        self._id = -1
-        self._insertion = Vector(0, 0, 1)
+        self._obb = None
+        self._collision_mesh = None
+        self._dimensions = []
+        self._features = {}
+        self._insertion = Vector(0, 0, -1)
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: why is the default the negative z-axis?

File: src/compas_model/elements/element.py [hunk 12]
```diff
@@ -2,87 +2,320 @@
...
+        self._node = None
+
+    # ==========================================================================
+    # Serialization.
+    # ==========================================================================
+
+    @property
+    def data(self):
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: this should now be `__data__`

File: src/compas_model/elements/element.py [hunk 13]
```diff
@@ -2,87 +2,320 @@
...
+            "dimensions": self.dimensions,
+            "features": self.features,
+            "insertion": self.insertion,
+            "attributes": self.attributes,
+        }
+
+        return data
+
+    @classmethod
+    def from_data(cls, data):
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: this should now be `__from_data__`

File: src/compas_model/elements/element.py [hunk 14]
```diff
@@ -2,87 +2,320 @@
...
+    def collision_mesh(self):
+        if not self._collision_mesh:
+            self.compute_collision_mesh()
+        return self._collision_mesh
+
+    @property
+    def dimensions(self):
+        if not type(self.obb) is Box:
+            raise TypeError("OBB is not set as a Box.")
+        return [self.obb.width, self.obb.height, self.obb.depth]
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: all read-only => mention in attributes overview

File: src/compas_model/elements/element.py [hunk 15]
```diff
@@ -2,87 +2,320 @@
...
+    @abstractmethod
+    def compute_collision_mesh(self):
+        """Computes the collision geometry of the element.
+
+        Returns
+        -------
+        :class:`compas.datastructures.Mesh`
+            The collision geometry of the element.
+        """
+
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py:246]: remove blank line

File: src/compas_model/elements/element.py [hunk 16]
```diff
@@ -2,87 +2,320 @@
...
+        if isinstance(geometries, list):
+            copied_geometries = []
+            for g in geometries:
+                copied_geometries.append(g)
+            return copied_geometries
+        else:
+            return geometries.copy()
+
+    # ==========================================================================
+    # Python provided methods for printing and etc.
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: these should be listed towards the top

File: src/compas_model/elements/element.py [hunk 17]
```diff
@@ -2,87 +2,320 @@
...
+        :class:`compas_model.elements.Element`
+            A new instance of the Element with the specified transformation applied.
+
+        """
+        new_instance = self.copy()
+        new_instance.transform(transformation)
+        return new_instance
+
+    # ==========================================================================
+    # Private methods.
```
Comment (tomvanmele) [inline src/compas_model/elements/element.py]: would list these at the top of the class def

File: src/compas_model/elements/element.py (hunk 4 above)
Comment (petrasvestartas) [reply src/compas_model/elements/element.py]: Added clarification:

```
    geometry_simplified : list[Any] or Any
        A simplified geometrical shape, e.g. a central Line for a beam, a polygon for a plate.
    geometry : list[Any] or Any
        Closed shape(s) representing the actual geometry of the element, e.g. a mesh for a block, a box for a beam.
```
    
Before I thought it should be a list, but after several classes implementation, works for both cases.

File: src/compas_model/elements/element.py (hunk 5 above)
Comment (petrasvestartas) [reply src/compas_model/elements/element.py]: Corrected.

File: src/compas_model/elements/element.py (hunk 6 above)
Comment (petrasvestartas) [reply src/compas_model/elements/element.py]: Added following explanation:

        These are custom geometrical objects added to the elements through operations made by the user.
        For example, a cutting shape for boolean difference operations, text identifiers.

File: src/compas_model/elements/element.py (hunk 9 above)
Comment (petrasvestartas) [reply src/compas_model/elements/element.py]: Switched in all the classes and gave an explanation example.

File: src/compas_model/elements/element.py (hunk 10 above)
Comment (petrasvestartas) [reply src/compas_model/elements/element.py]: corrected

File: src/compas_model/elements/element.py (hunk 11 above)
Comment (petrasvestartas) [reply src/compas_model/elements/element.py]: To represent a vertical insertion downwards. I added this explanation to docstrings.
Often this is used for translation animations. Can be also (0, 0, 1)

**Review by tomvanmele — APPROVED** [review]

**Pull request merged**